src/osemgrep/language_server/custom_requests/Search.ml

```diff
@@ -156,23 +169,83 @@ let filter_by_gitignore (files : Fpath.t list)
 (* Information gathering *)
 (*****************************************************************************)
 
-let mk_env (server : RPC_server.t) params =
+let mk_env (server : RPC_server.t) (params : Request_params.t) =
   let scanning_roots =
     List_.map Scanning_root.of_fpath server.session.workspace_folders
   in
   let files =
     server.session.cached_workspace_targets |> Hashtbl.to_seq_values
     |> List.of_seq |> List.concat
   in
+  let project_root =
+    match
+      List.nth scanning_roots 0 |> Scanning_root.to_fpath |> Rfpath.of_fpath
+    with
+    | Error _ -> failwith "somehow unable to get project root from first root"
+    | Ok rfpath -> rfpath
+  in
   (* Filter all files by gitignores *)
   let filtered_files =
     (* Gitignore code. We hard-code it to be false for now.
        This adds 10s to startup time in `semgrep-app` on my machine.
        Gitignore is expensive! I don't know why.
     *)
-    if false then filter_by_gitignore files scanning_roots else files
+    if false then filter_by_gitignore ~project_root files else files
+  in
+  (* TODO: This has a bug!!!
+     Suppose we exclude `test.py` and are given `tests2/test.py`.
+     This code will not exclude properly, on the basis that `test.py`
+     does not match `tests2/test.py`.
+     Essentially, we may need to look at every suffix of the file to see
+     if it matches.
+  *)
+  let filtered_by_includes_excludes =
+    filtered_files
+    |> List.filter (fun file ->
+           (* Why must we do this?
+               The paths that we receive are absolute paths in the machine.
+               This means something like /Users/brandonspark/test/test.py.
+               When we match it against a blob, like `test.py`, obviously this
+               will not match, because of the giant absolute prefix.
+               We actually want the path _relative to the project root_, which is
+               `test.py` for the project `test`.
+               So we use `Fpath.rem_prefix` here, which emulates that functionality.
+           *)
```

Comment:
we should either assume the user will write *test.py for all test.py files, and we should not do this. Or if we assume they wont, we should auto add * to the pattern